### ISSUE TITLE
experimental/API pixel classification training/composition: annotations

### ISSUE THREAD

**Issue by k-dominik** [ilastik/ilastik#2397]
Right now the API supports prediction from pre-trained projects. In order to be more interesting to use in context of e.g. #2308, we need to be able to interactively change workflow parameters:

* selected feature set
  * also, query selected feature set
 * annotations to extract feature vectors to train on
  * need to agree on sparse format

This issue is to focus the discussion on a minimal specification for annotations (some uint array where pixel values correspond to label classes for data to be extracted from some kind of input in these positions).

I'm opening this issue in order to discuss possibilities that can be mapped to our current back-end, but would also play nicely with e.g. the backend of web-ilastik https://github.com/ilastik/ilastik/blob/API .

The two ideas for _annotations_ that we currently have floating around are 
1) the ilastik ones, where we have axis aligned, anchored (by roi) vigra arrays that will be cut up (if necessary) into a blocked array cache.
2) the web-ilastik ones, where we have numpy arrays (with axis specs), that are also anchored by a position.

I would say that we don't need to get into detail of how and where features are extracted. This should be implementation detail of the backend. However what we should agree on is:

1) What format should we settle on, blocks of `xarray.DataArray`, `ndstructs.Array5D`, 'vigra.Array`?
1) should the API save state -> save a list of annotations? Or, should annotations be uniquely identifiable in order to have some way of caching? 
1) Do we allow overlapping annotations? How do we deal with those?

cc @Tomaz-Vieira @m-novikov @emilmelnikov

**Comment by emilmelnikov** [comment]
> The two ideas for _annotations_ that we currently have floating around are
> 
> 1. the ilastik ones, where we have axis aligned, anchored (by roi) vigra arrays that will be cut up (if necessary) into a blocked array cache.
> 2. the web-ilastik ones, where we have numpy arrays (with axis specs), that are also anchored by a position.

I'd also suggest [`sparse.COO`](https://sparse.pydata.org/en/stable/), possibly wrapped into `xarray.DataArray`, together with the attached label metadata, stored in `.attrs`, something like `{1: {'name': 'Dead Cells', color: '#ff0000'}, 2: {'name': 'Live Cells', 'color': '#abcdef'}}`.

> 1. What format should we settle on, blocks of `xarray.DataArray`, `ndstructs.Array5D`, `vigra.Array`?

I vote for `xarray.DataArray` since it is already widely known in the numerical Python ecosystem.

> 2. should the API save state -> save a list of annotations? Or, should annotations be uniquely identifiable in order to have some way of caching?

Can't we just store all annotations in a single place? They are sparse, so size shouldn't be an issue, especially if we'll use `sparse.COO`. Undo (if we get to it someday) can be implemented just by keeping history directly as a list of sparse labels; this will require ~2x of the original sparse label size, but this shouldn't be an issue.

> 3. Do we allow overlapping annotations? How do we deal with those?

[Two `sparse.COO` arrays (old combined annotations, and a new annotation batch) can be (somewhat) efficiently merged](https://github.com/pydata/sparse/issues/394).

**Comment by Tomaz-Vieira** [comment]
Here's my biased perspective, based on the experience with webilastik's [Annotation](https://github.com/ilastik/webilastik/blob/bfec397a62b2ebc9a2df8df3744bffbdec3adc7c/webilastik/annotations/annotation.py#L90):

> 1. What format should we settle on, blocks of xarray.DataArray, ndstructs.Array5D, vigra.Array? 

I think any sparse array-like thing would be good enough, but I don't think it should be _just_ the raw array. Our annotations should be self-sufficient, that is, they must know what they represent (which axis is which, and what is the raw data that is being annotated by this Annotation), instead of having its interpretation reliant on the state of the workflow. Because of that, webilastik's annotations are arrays with labeled axes, and they also keep track of the `DataSource` onto which they apply.

> 2. should the API save state -> save a list of annotations? Or, should annotations be uniquely identifiable in order to have some way of caching?

This brings us back to that whole discussion of "programmatic API" vs "interactive API".

If we're talking just about the programmatic API, that is, that which:
 - interacts with libraries and scripts;
 - handles with `Operators`,
 
then no, the API should not save any state, and should require everything it needs (e.g.: annotations, features names) from the client code, every time.

If we're talking about the interactive API, that is, that which:
 -  interacts with the GUI;
 - is composed of applets and workflows;
 - expects humans to interactively set values;
 - produces `.ilp` files
 
 then yes, we should (must, even) keep state.

> 3. Do we allow overlapping annotations? How do we deal with those?

We should allow it, with later annotations overriding the common pixels from earlier annotations. Implementing undo/redo and snapping to annotations is extremely convenient if each annotation is its own object. The implementation must resolve those conflicts/overwrites before doing training, but that should should not influence the API

**Comment by k-dominik** [comment]
Ah, I knew this would be going to be a great discussion!

Let's maybe first clear up what our experimental api is for (at least my idea of it): I always have seen this API to be _high level_, for users to _interactively (by coding)_ do prediction/composition of projects, loading, and saving of project files.

I don't exactly see the distinction as @Tomaz-Vieira describes here:

> If we're talking just about the programmatic API, that is, that which:
> 
>     * interacts with libraries and scripts;
> 
>     * handles with `Operators`,
> 

and with this you mean the whole workflow - right, that is assembled of it's parts (so multiple chained computational operations that can be customized with some parameters). So no addition of annotations, what's done is done, and so on. Basically, the immutable workflow. (Although it still holds the state...). The resulting workflow object can do only one thing: predict with new data sources.

> then no, the API should not save any state, and should require everything it needs (e.g.: annotations, features names) from the client code, every time.
> 
> If we're talking about the interactive API, that is, that which:
> 
>     * interacts with the GUI;
> 
>     * is composed of applets and workflows;
> 
>     * expects humans to interactively set values;
> 
>     * produces `.ilp` files
> 
> 
> then yes, we should (must, even) keep state.
> 

So the second one I take is the _mutable_ workflow. You could potentially add annotations, change feature computations, all retraining/recomputation is handled transparently. This also needs to save some state of course.

For the annotation format, I @emilmelnikov's notion for sparse coo offers an efficient way without having to worry about offsets ourselves, which is nice. Also attaching meta data is nice, and, as @Tomaz-Vieira has mentioned, we should probably add the data_source there too. NGFF only has a [definition for (segmentation)label images](https://ngff.openmicroscopy.org/latest/#label-md) which is not sparse, but we could use a similar schema.


> Can't we just store all annotations in a single place? They are sparse, so size shouldn't be an issue, especially if we'll use `sparse.COO`. Undo (if we get to it someday) can be implemented just by keeping history directly as a list of sparse labels; this will require ~2x of the original sparse label size, but this shouldn't be an issue.

I agree with this. When it comes to undo, however, I'd say it's a gui functionality. But we need to discuss this more thoroughly.

**Comment by Tomaz-Vieira** [comment]
> and with this you mean the whole workflow - right, that is assembled of it's parts (so multiple chained computational operations that can be customized with some parameters). So no addition of annotations, what's done is done, and so on. Basically, the immutable workflow. (Although it still holds the state...). The rgh level" esulting workflow object can do only one thing: predict with new data sources.

Yes, that is exactly what I meant, and I'll adhere to your proposed terminology of "Mutable/Immutable Workflow".

> I always have seen this API to be high level, (...)

I think "high level API" might be a misnomer, though, because the concepts with which the programmatic API has to deal are the same as those with which the interactive API has to deal: features, annotations, data sources, predictions. So there is no further abstraction gained by having the API be interactive (or mutable, as you put it) 

> (...) for users to interactively (by coding) do prediction/composition of projects, loading, and saving of project files.

So this (interactively by coding) I find to be counter-productive. It is fine to have a user interface which is not graphical, like a text-mode pixel classification workflow or whatever. But if you're actually writing code, then it is detrimental to use the interactive (mutable) API.

If you write code that looks like this:

```python3
workflow = PixelWorkflow(...)
```

**Comment by Tomaz-Vieira** [comment]
> and with this you mean the whole workflow - right, that is assembled of it's parts (so multiple chained computational operations that can be customized with some parameters). So no addition of annotations, what's done is done, and so on. Basically, the immutable workflow. (Although it still holds the state...). The rgh level" esulting workflow object can do only one thing: predict with new data sources.

Yes, that is exactly what I meant, and I'll adhere to your proposed terminology of "Mutable/Immutable Workflow".

> I always have seen this API to be high level, (...)

I think "high level API" might be a misnomer, though, because the concepts with which the programmatic API has to deal are the same as those with which the interactive API has to deal: features, annotations, data sources, predictions. So there is no further abstraction gained by having the API be interactive (or mutable, as you put it) 

> (...) for users to interactively (by coding) do prediction/composition of projects, loading, and saving of project files.

So this ("interactively by coding") I find to be counter-productive. It is fine to have a user interface which is not graphical, like a text-mode pixel classification workflow or whatever. But if you're actually writing code, then it is detrimental to use the interactive (mutable) API.

If you write code that looks like this, in the mutable-style API:

```python3
workflow = PixelWorkflow(...)
workflow.predict(datasource)
```
the code doesn't look wrong, but you'll probably get a "MissingAnnotationsException" or "MissingFeaturesException" or whatever, and there isn't anything in the API that immediately tells you that you needed to add those. You have to look through all of the API, find all of the `addAnnotation`, `addFeature`, `addWhatever` methods, call those, then call the `predict` method, hoping that your workflow is now correctly set up. The answer to the question _"Am I ready to predict now?"_ is always _"maybe"_.

This behavior is acceptable in an interactive mode, because:

1 - The human users can't just dump all the necessary information the workflows needs all at once;
2 - The user interface will guide the user into inputting everything that the workflow needs;
3 - The UI will in fact prohibit the user from asking for predictions before the workflow is set up.

 What this ends up doing is that "coding interactively" means writing some code, running it, reading the hopefully useful exception message, then going back to your code, editing it, and trying again. This would be completely avoided if the workflow asked for everything that it needed clearly and explicitly (i.e.: in the workflow constructor or as parameters to the `predict` method):
 
 1 -  if you're using static type checking, you'll immediately be told that you're missing some stuff;
 2 - If you're not using static checking, you'll get something like `TypeError: missing required argument: annotations`, thrown automatically by python instead of manually by us;
 3 - This might not even happen, because the users will look at the API, see that it has a bunch of required arguments, and supply all of them;
 4 - And finally, when this kind of error happens, it is the _user's_ fault, for not reading the very clear API, not _ours_, for having an obscure API that requires some methods to be called before others.
